pkg/controller/networkpolicy/networkpolicy.go

```diff
@@ -305,3 +296,111 @@ func NetworkPolicyForService(req router.Request, resp router.Response) error {
+// It also only allows outgoing traffic to CoreDNS, the Acorn registry, and the Internet.
+func PolicyForBuilder(req router.Request, resp router.Response) error {
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return err
+	} else if !*cfg.NetworkPolicies {
+		return nil
+	}
+
+	deployment := req.Object.(*appsv1.Deployment)
+	if deployment.Name != "buildkitd" && !strings.HasPrefix(deployment.Name, "bld-") {
+		// this is not a builder deployment
+		return nil
+	}
+
+	podLabels := deployment.Spec.Template.ObjectMeta.Labels
+	externalIPBlock, err := buildExternalIPBlock(req)
+	if err != nil {
+		return err
+	}
+
+	// build the NetPol
+	resp.Objects(&networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deployment.Name,
+			Namespace: deployment.Namespace,
+			Labels: map[string]string{
+				labels.AcornManaged: "true",
+			},
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: podLabels,
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{{
+				From: []networkingv1.NetworkPolicyPeer{{
+					// allow access from the acorn-apiserver
+					NamespaceSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"kubernetes.io/metadata.name": system.Namespace,
+						},
+					},
+					PodSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"app": system.APIServerName,
+						},
+					},
+				}},
+			}},
+			Egress: []networkingv1.NetworkPolicyEgressRule{{
+				To: []networkingv1.NetworkPolicyPeer{
+					{
+						IPBlock: externalIPBlock, // allow traffic to the Internet
+					},
+					{
+						NamespaceSelector: &metav1.LabelSelector{
+							// allow traffic to kube-system for CoreDNS
+							MatchLabels: map[string]string{
+								"kubernetes.io/metadata.name": "kube-system",
+							},
+						},
+						PodSelector: &metav1.LabelSelector{
+							MatchLabels: map[string]string{
+								"k8s-app": "kube-dns",
+							},
+						},
+					},
+					{
+						PodSelector: &metav1.LabelSelector{
+							// allow traffic to the registry for pushing images
+							MatchLabels: map[string]string{
+								"app": system.RegistryName,
+							},
+						},
+					},
+				},
+			}},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress, networkingv1.PolicyTypeEgress},
+		},
+	})
+
+	return nil
+}
+
+// buildExternalIPBlock creates a NetworkPolicy IPBlock with the CIDR set to 0.0.0.0/0
+// and the Except set to the pod CIDRs of the nodes.
+func buildExternalIPBlock(req router.Request) (*networkingv1.IPBlock, error) {
+	ipBlock := networkingv1.IPBlock{
+		CIDR: "0.0.0.0/0",
+	}
+
+	// get pod CIDRs from the nodes so that we can only allow IP addresses outside the cluster
+	nodes := corev1.NodeList{}
+	if err := req.Client.List(req.Ctx, &nodes); err != nil {
+		return nil, fmt.Errorf("failed to list nodes: %w", err)
+	}
+	for _, node := range nodes.Items {
+		for _, cidr := range node.Spec.PodCIDRs {
```

Comment:
Should it block service IP cidr too?

Reply:
I definitely think we should do this, but the trouble is that I can't find a way to get this information programatically. Apparently it is sometimes in the output of `kubectl cluster-info dump` but it did not appear there for me. We could make this another installation parameter that the user can optionally provide if you think that would be good.

Reply:
yeah it is true. What was the original intention here? To block buildkit pod being accessed from other buildkit?

Reply:
This is what @cjellick wanted originally:

"Builders all go in the same namespace. They aren't isolated form each other. Should do a policy specifically for builders. Isolate them from each other as though they're in their own projects. Should only be able to talk to registry and internet"

Based on that last sentence I decided to restrict egress traffic. I think this makes sense to do, since users can run whatever commands they want to run in their Dockerfiles, which are also (potentially) built during the Acorn build process.

I wish I could set an explicit deny to all traffic in all namespaces by default, and then use the ipBlock to let it get out to the Internet. But I'm not sure if there is a way to do that.

Reply:
Got it. Just curious, with your change, can you still access other buildkit service using service IP? I wonder when kube-proxy translate service IP into pod ip, the network policy will still block the pod ip, then we don't have to worry about it.

Reply:
Just tested this in a local cluster. Thankfully, just using the pod IP CIDR works just fine. I deployed a pod in the default namespace and exposed with with a service. Without this NetPol, I was able to curl that pod just fine, by both its service hostname (`nginx.default.svc.cluster.local`) and by the Service IP that that hostname resolves to. When I added the NetworkPolicy to restrict egress traffic, I could no longer reach that service, by its hostname or its IP. So I think we are good here. 👍